src/sentry/api/serializers/rest_framework/sentry_app.py

```diff
@@ -86,7 +86,7 @@ def validate_name(self, value):
         return value
 
     def validate(self, attrs):
-        if not attrs.get('scopes'):
+        if not attrs.get('scopes') or not attrs.get('events'):
```

Comment:
So I think what could happen here is someone updates their permissions (aka scopes) but doesn't update their events and then we wouldn't check that they still have the appropriate permissions for the previous events. 

**Scenario:**
* App has scopes `['event:read']` and events `[u'issue.assigned', u'issue.resolved', u'issue.ignored', u'issue.created']`

* User updates scopes to be `['project:read']` but doesn't send along 'events' in the payload

* App then has scopes `['project:read']` but then also events `[u'issue.assigned', u'issue.resolved', u'issue.ignored', u'issue.created']`

Reply:
Does this mean that events should be always `required` then?

Reply:
Well that seems a bit annoying to have to send 'events' if you are just updating the author or something like that.

Reply:
So `events` is basically required if `scopes` is updated (and vice-versa)?

Reply:
Actually, I think this case could have been triggered before. If you called the API with no `events` but updated the `scopes` then I don't think it would have triggered `validate_events`, and so the scope change would have gone through regardless.

It seems like we just want to enforce that both `events` and `scopes` were passed if either was passed?

Reply:
Ideally we just want to make sure the result of a change to either one, is still valid when considering both. I don't think requiring both to be passed, when either changes, is the right experience, personally.

In my opinion, the way this should work is basically:

1. get new state of `events` and `scopes` (using either the values passed in or the existing values from the object)
2. validate the required scope is present for each event


```python
scopes = attrs.get('scopes', sentry_app.scopes)
events = attrs.get('events', sentry_app.events)

for event in events:
    if not has_scope_required_for_event(event, scopes):
        raise ValidationError
```

Reply:
> It seems like we just want to enforce that both `events` and `scopes` were passed if either was passed?

yeah I think this sounds about right

Reply:
can this be done here https://github.com/getsentry/sentry/blob/master/src/sentry/mediators/sentry_apps/updater.py#L61 since we also already do a check in `_update_events`

Reply:
> can this be done here https://github.com/getsentry/sentry/blob/master/src/sentry/mediators/sentry_apps/updater.py#L61 since we also already do a check in `_update_events`

Was about to suggest this as well, it seems like the mediators might be the right place for this  since we actually have access to the instance at that point. Since we're not using a `ModelSerializer` here we can't easily perform these comparisons.

Reply:
Does Django have some notion of model validations? (opposed to pure input value validations that the "serializers" seem to support).

Reply:
DRF has `ModelSerializers` that operate on an actual django model instance. We use them in a couple of places, but almost entirely use them differently than intended. `ApiKeySerializer` seems to be the best example of us using one to actually update an instance, although it doesn't have any additional validation on it.

https://www.django-rest-framework.org/api-guide/serializers/#modelserializer